openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/strcompat/doc/man/man9/aixcompat.9.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) aixcompat.9.man,v 0.9.2.8 2006/04/05 09:43:30 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:43:30 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" aixcompat.9.man,v
.\" Revision 0.9.2.8  2006/04/05 09:43:30  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database strcompat.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so strcompat.macros
.\"
.\"
.TH AIXCOMPAT 9 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ DDI/DKI"
.\"
.\"
.SH NAME
.B aixcompat
\- AIX \fISTREAMS\fP Compatibility
.\"
.\"
.SH SYNOPSIS
.PP
.B #define _AIX_SOURCE
.br
.B #include <sys/ddi.h>
.\"
.\"
.SH OVERVIEW
.PP
.B @PACKAGE_TITLE@
provides a compatibility module,
.I streams-aixcompat
that provides compatibility functions that smooth porting of drivers from the
.I AIX 5L Version 5.1
.I STREAMS
environment to
.BR "@PACKAGE_TITLE@" .
The functions provided by this compatibility module are described here.
.\"
.\"
.SH DESCRIPTION
.PP
Many of the functions that are provided by
.IR AIX \(rg
are also provided by the core
.I streams
kernel module of
.BR "@PACKAGE_TITLE@" .
Other
.IR AIX \(rg-specific
functions are provided by the
.I streams-aixcompat
compatibility kernel module.
.\"
.\"
.SS "Kernel Functions"
.PP
Kernel functions available in
.IR AIX \(rg
that are already available in
.B Linux
are as follows:
.HP
.BR atoi (3)\ \-\ \c
convert a string to an integer
.PD 0
.HP
.BR bcmp (3)\ \-\ \c
compare byte strings
.HP
.BR bcopy (3)\ \-\ \c
copy byte strings
.HP
.BR bzero (3)\ \-\ \c
write zeros to a byte string
.HP
.BR lbolt (9)\ \-\ \c
time in ticks since reboot
.HP
.BR memccpy (3)\ \-\ \c
copy memory area
.HP
.BR memchr (3)\ \-\ \c
scan memory for a character
.HP
.BR memcmp (3)\ \-\ \c
compare memory areas
.HP
.BR memcpy (3)\ \-\ \c
copy memory area
.HP
.BR memmove (3)\ \-\ \c
copy memory area
.HP
.BR memset (3)\ \-\ \c
fill memory with a constant byte
.HP
.BR ovbcopy (3)\ \-\ \c
overlapping byte copy
.HP
.BR splx (9)\ \-\ \c
restore priority levels
.HP
.BR strcat (3)\ \-\ \c
concatenate two strings
.HP
.BR strchr (3)\ \-\ \c
locate character in string
.HP
.BR strcmp (3)\ \-\ \c
compare two strings
.HP
.BR strcspn (3)\ \-\ \c
search a string for a set of characters
.HP
.BR strlen (3)\ \-\ \c
calculate the length of a string
.HP
.BR strncat (3)\ \-\ \c
concatenate two strings
.HP
.BR strncmp (3)\ \-\ \c
compare two strings
.HP
.BR strncpy (3)\ \-\ \c
copy a string
.HP
.BR strpbrk (3)\ \-\ \c
search a string for any of a set of characters
.HP
.BR strrchr (3)\ \-\ \c
locate character in string
.HP
.BR strspn (3)\ \-\ \c
search a string for a set of characters
.HP
.BR strstr (3)\ \-\ \c
locate a substring
.HP
.BR strtok (3)\ \-\ \c
extract tokens from strings
.HP
.BR time (9)\ \-\ \c
time in seconds since epoch
.PD
.\"
.SS "\fISTREAMS\fP Functions"
.PP
.I STREAMS
functions that are already available in
.B @PACKAGE_TITLE@
(in the core
.I streams
kernel module) are as follows:
.HP
.BR adjmsg (9)\ \-\ \c
trim bytes from the front or back of a \fISTREAMS\fP message
.PD 0
.HP
.BR allocb (9)\ \-\ \c
allocate a \fISTREAMS\fP message and data block
.HP
.BR backq (9)\ \-\ \c
find the upstream or downstream queue
.HP
.BR bcanput (9)\ \-\ \c
test flow control on a \fISTREAMS\fP message queue
.HP
.BR bufcall (9)\ \-\ \c
install a buffer callback
.HP
.BR canput (9)\ \-\ \c
test flow control on a message queue
.HP
.BR copyb (9)\ \-\ \c
copy a \fISTREAMS\fP message block
.HP
.BR copymsg (9)\ \-\ \c
copy a \fISTREAMS\fP message
.HP
.BR datamsg (9)\ \-\ \c
tests a \fISTREAMS\fP message type for data
.HP
.BR dupb (9)\ \-\ \c
duplicate a \fISTREAMS\fP message block
.HP
.BR dupmsg (9)\ \-\ \c
duplicate a \fISTREAMS\fP message
.HP
.BR enableok (9)\ \-\ \c
allows a \fISTREAMS\fP message queue to be scheduled
.HP
.BR esballoc (9)\ \-\ \c
allocate a \fISTREAMS\fP message and data block with a caller supplied data buffer
.HP
.BR flushband (9)\ \-\ \c
flushes band \fISTREAMS\fP messages from a message queue
.HP
.BR flushq (9)\ \-\ \c
flushes messages from a \fISTREAMS\fP message queue
.HP
.BR freeb (9)\ \-\ \c
frees a \fISTREAMS\fP message block
.HP
.BR freemsg (9)\ \-\ \c
frees a \fISTREAMS\fP message
.HP
.BR getadmin (9)\ \-\ \c
get the administration function pointer for a \fISTREAMS\fP module
.HP
.BR getmid (9)\ \-\ \c
get the \fISTREAMS\fP module id for a name
.HP
.BR getq (9)\ \-\ \c
gets a message from a \fISTREAMS\fP message queue
.HP
.BR insq (9)\ \-\ \c
inserts a message into a \fISTREAMS\fP message queue
.HP
.BR linkb (9)\ \-\ \c
link a message block to a \fISTREAMS\fP message
.HP
.BR msgdsize (9)\ \-\ \c
calculate the size of the data in a \fISTREAMS\fP message
.HP
.BR noenable (9)\ \-\ \c
disables a \fISTREAMS\fP message queue from being scheduled
.HP
.BR OTHERQ (9)\ \-\ \c
return the other queue of a \fISTREAMS\fP queue pair
.HP
.BR pullupmsg (9)\ \-\ \c
pull up the bytes in a \fISTREAMS\fP message
.HP
.BR putbq (9)\ \-\ \c
put a message back on a \fISTREAMS\fP message queue
.HP
.BR putctl1 (9)\ \-\ \c
put a 1 byte control message on a \fISTREAMS\fP message queue
.HP
.BR putctl (9)\ \-\ \c
put a control message on a \fISTREAMS\fP message queue
.HP
.BR putnext (9)\ \-\ \c
put a message on the downstream \fISTREAMS\fP message queue
.HP
.BR putq (9)\ \-\ \c
put a message on a \fISTREAMS\fP message queue
.HP
.BR qenable (9)\ \-\ \c
schedules a \fISTREAMS\fP message queue service routine
.HP
.BR qreply (9)\ \-\ \c
replys to a message from a \fISTREAMS\fP message queue
.HP
.BR qsize (9)\ \-\ \c
return the number of messages on a queue
.HP
.BR RD (9)\ \-\ \c
return the read queue of a \fISTREAMS\fP queue pair
.HP
.BR rmvb (9)\ \-\ \c
remove a message block from a \fISTREAMS\fP message
.HP
.BR rmvq (9)\ \-\ \c
remove a message from a \fISTREAMS\fP message queue
.HP
.BR splstr (9)\ \-\ \c
set priority levels
.HP
.BR splx (9)\ \-\ \c
restore priority levels
.HP
.BR strlog (9)\ \-\ \c
pass a message to the \fISTREAMS\fP logger
.HP
.BR strqget (9)\ \-\ \c
gets information about a \fISTREAMS\fP message queue
.HP
.BR testb (9)\ \-\ \c
test if a \fISTREAMS\fP message can be allocated
.HP
.BR timeout (9)\ \-\ \c
start a timer
.HP
.BR unbufcall (9)\ \-\ \c
removes a \fISTREAMS\fP buffer callback
.HP
.BR unlinkb (9)\ \-\ \c
unlink a message block from a \fISTREAMS\fP message
.HP
.BR untimeout (9)\ \-\ \c
stop a timer
.HP
.BR unweldq (9)\ \-\ \c
unweld two queues
.HP
.BR weldq (9)\ \-\ \c
weld two (or four) queues together
.HP
.BR WR (9)\ \-\ \c
return the write queue of a \fISTREAMS\fP queue pair
.PD
.\"
.\"
.SS "\fIMPS\fP\fR\(rg\fP Functions"
.PP
.IR AIX \(rg
exposes a number of symbols from the underlying Mentat
.IR MPS \(rg
implementation that are also documented in the
.IR AIX \(rg
documentation.
.I STREAMS
functions that are available to
.IR AIX \(rg
in the
.B @PACKAGE_TITLE@
.IR MPS \(rg
compatibility module,
.IR streams-mpscompat ,
are as follows:
.HP
.BR mi_bufcall (9)\ \-\ \c
reliable alternative to \fBbufcall\fP(9)
.PD 0
.HP
.BR mi_close_comm (9)\ \-\ \c
\fISTREAMS\fP common minor device close utility
.HP
.BR mi_next_ptr (9)\ \-\ \c
\fISTREAMS\fP minor device list traversal
.HP
.BR mi_open_comm (9)\ \-\ \c
\fISTREAMS\fP common minor device open utility
.HP
.BR mi_prev_ptr (9)\ \-\ \c
\fISTREAMS\fP minor device list traversal
.PD
.\"
.\"
.SS "Specific \fISTREAMS\fP Functions"
.PP
.I STREAMS
functions that are available in the
.B @PACKAGE_TITLE@
.IR AIX \(rg
compatibility module,
.IR streams-aixcompat ,
are as follows:
.HP
.BR wantio (9)\ \-\ \c
perform direct I/O from a \fISTREAMS\fP driver
.PD 0
.HP
.BR wantmsg (9)\ \-\ \c
provide a filter of wanted message for a \fISTREAMS\fP module
.PD
.\"
.\"
.SS "\fISTREAMS\fP Configuration and Registration Functions"
.PP
Module and driver configuration functions that are available in the
.B @PACKAGE_TITLE@
.IR AIX \(rg
compatibility module,
.IR streams-aixcompat ,
are as follows:
.HP
.BR str_install (9)\ \-\ \c
install or uninstall a \fISTREAMS\fP module or driver
.PD 0
.HP
.BR str_install_AIX (9)\ \-\ \c
install or uninstall a \fISTREAMS\fP module or driver
.PD
.\"
.\"
.SH USAGE
.PP
For usage an example on the various functions, see the individual manual pages.
.\"
.\"
.SH CONTEXT
.PP
These functions can be executed from various contexts as described in each separate manual page.
.\"
.\"
.SH MP-STREAMS
.PP
Most of these functions are MP-safe; however,
.IR AIX \(rg
does not provide many of the
.I SVR 4.2 MP
functions, such as:
.HP
.BR bcanputnext (9)\ \-\ \c
test flow control on a \fISTREAMS\fP message queue
.PD 0
.HP
.BR canputnext (9)\ \-\ \c
test flow control on a \fISTREAMS\fP message queue
.HP
.BR esbbcall (9)\ \-\ \c
install a buffer callback for an extended \fISTREAMS\fR message block
.HP
.BR freezestr (9)\ \-\ \c
freeze the state of a stream queue
.HP
.BR msgpullup (9)\ \-\ \c
pull up bytes in a \fISTREAMS\fP message
.HP
.BR put (9)\ \-\ \c
invoke the put procedure for a \fISTREAMS\fP module or driver with a \fISTREAMS\fP message
.HP
.BR putnextctl1 (9)\ \-\ \c
put a 1 byte control message on the downstream \fISTREAMS\fR message queue
.HP
.BR putnextctl (9)\ \-\ \c
put a control message on the downstream \fISTREAMS\fR message queue
.HP
.BR qprocsoff (9)\ \-\ \c
disables \fISTREAMS\fR message queue processing for multi-processing
.HP
.BR qprocson (9)\ \-\ \c
enables a \fISTREAMS\fR message queue for multi-processing
.HP
.BR SAMESTR (9)\ \-\ \c
test for \fISTREAMS\fR pipe or fifo
.HP
.BR unfreezestr (9)\ \-\ \c
thaw the state of a stream queue
.PD
.PP
Lack of these functions can make Multiprocessor (MP) support problematic.
Ported
.IR AIX \(rg
.I STREAMS
modules and drivers should always be loaded under
.B @PACKAGE_TITLE@
with multiprocessing disabled for safety.
.\"
.\"
.SH NOTICES
.PP
.B _AIX_SOURCE
must be defined before including DDI/DKI include files to expose
.IR AIX \(rg-specific
symbols.
To use
.IR AIX \(rg-specific
functions,
.B @PACKAGE_TITLE@
must also have been compiled with
.B CONFIG_STREAMS_COMPAT_AIX
or
.B CONFIG_STREAMS_COMPAT_AIX_MODULE
defined.
When compiled with
.B CONFIG_STREAMS_COMPAT_AIX
defined,
.IR AIX \(rg
compatibility will become part of the
.B @PACKAGE_TITLE@
.I streams
kernel module.
When compiled with
.B CONFIG_STREAMS_COMPAT_AIX_MODULE
defined,
.IR AIX \(rg
compatibility will be placed in a separate
.I streams-aixcompat
kernel module that will demand load when any
.I STREAMS
driver or module that uses the
.IR AIX \(rg-specific
functions loads.
.PP
.IR MPS \(rg
functions used by
.IR AIX \(rg
are only be available from the
.I streams-mpscompat
compatibility module.  See
.BR mpscompat (9)
for more information on this module.
.\"
.\"
.SH IMPLEMENTATION
.PP
Core functions are implemented within the
.B @PACKAGE_TITLE@
framework.
.IR AIX \(rg-specific
functions are implemented as described in the separate manual pages describing each facility.
.\"
.\"
.SH "SEE ALSO"
.PP
Core functions:
.BR adjmsg (9),
.BR allocb (9),
.BR backq (9),
.BR bcanput (9),
.BR bufcall (9),
.BR canput (9),
.BR copyb (9),
.BR copymsg (9),
.BR datamsg (9),
.BR dupb (9),
.BR dupmsg (9),
.BR enableok (9),
.BR esballoc (9),
.BR flushband (9),
.BR flushq (9),
.BR freeb (9),
.BR freemsg (9),
.BR getadmin (9),
.BR getmid (9),
.BR getq (9),
.BR insq (9),
.BR linkb (9),
.BR msgdsize (9),
.BR noenable (9),
.BR OTHERQ (9),
.BR pullupmsg (9),
.BR putbq (9),
.BR putctl1 (9),
.BR putctl (9),
.BR putnext (9),
.BR putq (9),
.BR qenable (9),
.BR qreply (9),
.BR qsize (9),
.BR RD (9),
.BR rmvb (9),
.BR rmvq (9),
.BR splstr (9),
.BR splx (9),
.BR strlog (9),
.BR strqget (9),
.BR testb (9),
.BR timeout (9),
.BR unbufcall (9),
.BR unlinkb (9),
.BR untimeout (9),
.BR unweldq (9),
.BR weldq (9),
.BR WR (9).

.PP
Compatibility functions:
.BR mi_bufcall (9),
.BR mi_close_comm (9),
.BR mi_next_ptr (9),
.BR mi_open_comm (9),
.BR mi_prev_ptr (9),
.BR str_install (9),
.BR str_install_AIX (9),
.BR wantio (9),
.BR wantmsg (9).
.PP
General:
.BR STREAMS (9).
.\"
.\"
.SH BUGS
.PP
The
.IR AIX \(rg
compatibility suite has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
These functions are compatible with
.IR AIX \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
Neither
.IR AIX \(rg-specific,
nor
.IR MPS \(rg,
functions should be used by portable
.I STREAMS
drivers and modules.
The more common
.I SVR 4.2 MP
functions should be used in their stead.
.IP \(em
.IR AIX \(rg-specific
functions are provided for source level compatibility with
.IR AIX \(rg
for the purpose of porting
.IR AIX \(rg
drivers to
.BR Linux .
Because internal structures and definitions for
.B @PACKAGE_TITLE@
.[
lfs
.]
differ from that of
.IR AIX \(rg,
binaries are incompatible.
.IP \(em
.IR AIX \(rg-specific
functions only provide functionality comparable to
.IR AIX \(rg
for use by
.I STREAMS
drivers and modules.
Any and all functionality provided by
.IR AIX \(rg-specific
functions in support of
.RI non- STREAMS
devices are not supported.
.PP
See
.BR STREAMS (9)
for additional compatibility considerations.
.\"
.\"
.SH CONFORMANCE
.PP
.IR AIX \(rg
DDI documentation.
.[
aixdoc
.]
.[
aixspg
.]
.\"
.\"
.SH HISTORY
.PP
Most of these functions
first appeared in
.IR AIX \(rg;
except functions with a name starting with
.IR mi_ .
Those starting with
.I mi_
are
.IR MPS \(rg
functions dating back to
.RI pre- OSF/1
.I STREAMS
implementations.
.\"
.\"
.[
$LIST$
.]
.TI


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/strcompat/src/kernel/aixcompat.c
/*****************************************************************************

 @(#) aixcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.23) 2006/07/07 20:49:55

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2005  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/07/07 20:49:55 by brian

 *****************************************************************************/

#ident "@(#) aixcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.23) 2006/07/07 20:49:55"

static char const ident[] =
    "aixcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.23) 2006/07/07 20:49:55";

/* 
 *  This is my solution for those who don't want to inline GPL'ed functions or
 *  who don't use optimizations when compiling or specifies
 *  -fnoinline-functions or something of the like.  This file implements all
 *  of the extern inlines from the header files by just including the header
 *  files with the functions declared 'inline' instead of 'extern inline'.
 *
 *  There are implemented here in a separate object, out of the way of the
 *  modules that don't use them.
 */

#define __AIX_EXTERN_INLINE __inline__ streamscall

#define _AIX_SOURCE

#include "sys/os7/compat.h"

#define AIXCOMP_DESCRIP		"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define AIXCOMP_COPYRIGHT	"Copyright (c) 1997-2005 OpenSS7 Corporation.  All Rights Reserved."
#define AIXCOMP_REVISION	"LfS aixcompat.c,v OpenSS7-0_9_2_D_rc2(0.9.2.23) 2006/07/07 20:49:55"
#define AIXCOMP_DEVICE		"AIX 5L Version 5.1 Compatibility"
#define AIXCOMP_CONTACT		"Brian Bidulock <bidulock@openss7.org>"
#define AIXCOMP_LICENSE		"GPL"
#define AIXCOMP_BANNER		AIXCOMP_DESCRIP		"\n" \
				AIXCOMP_COPYRIGHT	"\n" \
				AIXCOMP_REVISION	"\n" \
				AIXCOMP_DEVICE		"\n" \
				AIXCOMP_CONTACT		"\n"
#define AIXCOMP_SPLASH		AIXCOMP_DEVICE		" - " \
				AIXCOMP_REVISION	"\n"

#ifdef CONFIG_STREAMS_COMPAT_AIX_MODULE
MODULE_AUTHOR(AIXCOMP_CONTACT);
MODULE_DESCRIPTION(AIXCOMP_DESCRIP);
MODULE_SUPPORTED_DEVICE(AIXCOMP_DEVICE);
MODULE_LICENSE(AIXCOMP_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-aixcompat");
#endif
#endif

/* 
 *  MI_OPEN_COMM, MI_CLOSE_COMM, MI_NEXT_PTR, MI_PREV_PTR and MI_BUFCALL
 *  -------------------------------------------------------------------------
 *  implemented in mspcompat.c
 */

#ifdef LFS
/* 
 *  WANTIO
 *  -------------------------------------------------------------------------
 */
int
wantio(queue_t *q, struct wantio *w)
{
	queue_t *wq = WR(q);
	struct stdata *sd = ((struct queinfo *) (wq - 1))->qu_str;

	if (w) {
		// qget(wq);
		sd->sd_directio = w;
	} else {
		sd->sd_directio = w;
		// qput(&wq);
	}
	return (0);
}

EXPORT_SYMBOL_NOVERS(wantio);		/* aix/ddi.h */

/* 
 *  WANTMSG
 *  -------------------------------------------------------------------------
 */
__AIX_EXTERN_INLINE int wantmsg(queue_t *q, int streamscall (*func) (mblk_t *));

EXPORT_SYMBOL_NOVERS(wantmsg);		/* aix/ddi.h */
#endif

/* 
 *  STR_INSTALL
 *  -------------------------------------------------------------------------
 */
int
str_install_AIX(int cmd, strconf_t * sc)
{
	if (!sc)
		return (EINVAL);
	switch (cmd) {
	case STR_LOAD_DEV:
#ifdef LIS
	{
		int err;

		if ((err = lis_register_strdev(sc->sc_major, sc->sc_str, 255, sc->sc_name)) > 0)
			sc->sc_major = err;
		return (err < 0 ? -err : 0);
	}
#endif
#ifdef LFS
		{
			struct cdevsw *cdev;
			int err;

#ifdef MAX_CHRDEV
			if (0 >= sc->sc_major || sc->sc_major >= MAX_CHRDEV)
#else
			if (sc->sc_major != MAJOR(MKDEV(sc->sc_major, 0)))
#endif
				return (EINVAL);
			/* We don't do old-style opens */
			if (!(sc->sc_open_stylesc_flags & STR_NEW_OPEN))
				return (ENOSYS);
			if (sc->sc_open_stylesc_flags & STR_Q_NOTTOSPEC) {
				/* Modules can always sleep because we are running at soft IRQ. */
			}
			if (!(cdev = kmem_zalloc(sizeof(*cdev), KM_NOSLEEP)))
				return (ENOMEM);
			cdev->d_name = sc->sc_name;
			cdev->d_str = sc->sc_str;
			/* build flags */
			cdev->d_flag = 0;
			if (sc->sc_open_stylesc_flags & STR_MPSAFE) {
				cdev->d_flag |= D_MP;
			}
			if (sc->sc_open_stylesc_flags & STR_QSAFETY) {
				cdev->d_flag |= D_SAFE;
			}
			if (sc->sc_open_stylesc_flags & STR_PERSTREAM) {
				cdev->d_flag |= D_UP;
			}
			if (sc->sc_open_stylesc_flags & STR_NEWCLONING) {
				cdev->d_flag |= D_CLONE;
			}
			switch ((cdev->d_sqlvl = sc->sc_sqlevel)) {
			case SQLVL_NOP:
				cdev->d_flag |= D_MP;
				break;
			case SQLVL_QUEUE:
				cdev->d_flag |= D_MTPERQ;
				break;
			case SQLVL_QUEUEPAIR:
				cdev->d_flag |= D_MTQPAIR;
				break;
			case SQLVL_MODULE:
				cdev->d_flag |= D_MTPERMOD;
				break;
			case SQLVL_ELSEWHERE:
				cdev->d_flag |= D_MTOUTPERIM;
				break;
			case SQLVL_GLOBAL:
				/* can't really support this, but its only used for debug anyway */
				cdev->d_flag &= ~D_MP;
				break;
			case SQLVL_DEFAULT:
				cdev->d_flag |= D_MTPERMOD;
				break;
			}
			if ((err = register_strdev(cdev, sc->sc_major)) < 0)
				kmem_free(cdev, sizeof(*cdev));
			return (-err);
		}
#endif
	case STR_UNLOAD_DEV:
#ifdef LIS
	{
		return lis_unregister_strdev(sc->sc_major);
	}
#endif
#ifdef LFS
		{
			struct cdevsw *cdev;
			int err;

			if (0 >= sc->sc_major || sc->sc_major >= MAX_STRDEV)
				return (EINVAL);
			if ((cdev = sdev_get(sc->sc_major)) == NULL)
				return (ENOENT);
			printd(("%s: %s: got device\n", __FUNCTION__, cdev->d_name));
			printd(("%s: %s: putting device\n", __FUNCTION__, cdev->d_name));
			sdev_put(cdev);
			if ((err = unregister_strdev(cdev, sc->sc_major)) == 0)
				kmem_free(cdev, sizeof(*cdev));
			return (-err);
		}
#endif
	case STR_LOAD_MOD:
#ifdef LIS
	{
		int err;

		if ((err = lis_register_strmod(sc->sc_str, sc->sc_name)) > 0)
			sc->sc_major = err;
		return (err < 0 ? -err : 0);
	}
#endif
#ifdef LFS
		{
			struct fmodsw *fmod;
			int err;

			if (!sc->sc_str || !sc->sc_str->st_rdinit
			    || !sc->sc_str->st_rdinit->qi_minfo)
				return (EINVAL);
			if (0 >= sc->sc_major || sc->sc_major >= MAX_STRDEV)
				return (EINVAL);
			/* We don't do old-style opens */
			if (!(sc->sc_open_stylesc_flags & STR_NEW_OPEN))
				return (ENOSYS);
			if (!(fmod = kmem_zalloc(sizeof(*fmod), KM_NOSLEEP)))
				return (ENOMEM);
			fmod->f_name = sc->sc_name;
			fmod->f_str = sc->sc_str;
			/* build flags */
			fmod->f_flag = 0;
			if (sc->sc_open_stylesc_flags & STR_MPSAFE) {
				fmod->f_flag |= D_MP;
			}
			if (sc->sc_open_stylesc_flags & STR_PERSTREAM) {
				fmod->f_flag |= D_UP;
			}
			if (sc->sc_open_stylesc_flags & STR_Q_NOTTOSPEC) {
				/* Modules can always sleep because we are running at soft IRQ;
				   however, it has a horrendous impact.  We just ignore it. */
			}
			if (sc->sc_open_stylesc_flags & STR_QSAFETY) {
				fmod->f_flag |= D_SAFE;
			}
			if (sc->sc_open_stylesc_flags & STR_NEWCLONING) {
				fmod->f_flag |= D_CLONE;
			}
			switch ((fmod->f_sqlvl = sc->sc_sqlevel)) {
			case SQLVL_NOP:
				fmod->f_flag |= D_MP;
				break;
			case SQLVL_QUEUE:
				fmod->f_flag |= D_MTPERQ;
				break;
			case SQLVL_QUEUEPAIR:
				fmod->f_flag |= D_MTQPAIR;
				break;
			case SQLVL_MODULE:
				fmod->f_flag |= D_MTPERMOD;
				break;
			case SQLVL_ELSEWHERE:
				fmod->f_flag |= D_MTOUTPERIM;
				break;
			case SQLVL_GLOBAL:
				/* can't really support this, but its only used for debug anyway */
				fmod->f_flag &= ~D_MP;
				break;
			case SQLVL_DEFAULT:
				fmod->f_flag |= D_MTPERMOD;
				break;
			}
			if ((err = register_strmod(fmod)) < 0)
				kmem_free(fmod, sizeof(*fmod));
			return (-err);
		}
#endif
	case STR_UNLOAD_MOD:
#ifdef LIS
	{
		return lis_unregister_strmod(sc->sc_str);
	}
#endif
#ifdef LFS
		{
			struct fmodsw *fmod;
			int err;

			if (!(fmod = fmod_str(sc->sc_str)))
				return (ENOENT);
			if ((err = unregister_strmod(fmod)) == 0)
				kmem_free(fmod, sizeof(fmod));
			return (-err);
		}
#endif
	}
	return (EINVAL);
}

EXPORT_SYMBOL_NOVERS(str_install_AIX);	/* strconf.h */

#ifdef CONFIG_STREAMS_COMPAT_AIX_MODULE
static
#endif
int __init
aixcomp_init(void)
{
#ifdef CONFIG_STREAMS_COMPAT_AIX_MODULE
	printk(KERN_INFO AIXCOMP_BANNER);
#else
	printk(KERN_INFO AIXCOMP_SPLASH);
#endif
	return (0);
}

#ifdef CONFIG_STREAMS_COMPAT_AIX_MODULE
static
#endif
void __exit
aixcomp_exit(void)
{
	return;
}

#ifdef CONFIG_STREAMS_COMPAT_AIX_MODULE
module_init(aixcomp_init);
module_exit(aixcomp_exit);
#endif
